File: utils/sql_validator.py
import re
from config import FORBIDDEN_KEYWORDS, DEFAULT_LIMIT

DANGEROUS_PATTERNS = [
    r'\bEXEC\b', r'\bEXECUTE\b',
    r'\bxp_\w+', r'\bsp_\w+',
    r'\bINTO\s+OUTFILE\b',
    r'\bINTO\s+DUMPFILE\b',
    r'\bLOAD_FILE\b',
    r'\bUNION\s+SELECT\b',
    r'\bSLEEP\s*\(',
    r'\bBENCHMARK\s*\(',
    r'\bWAITFOR\b',
    r'\bPG_SLEEP\b',
    r'\bATTACH\b',
    r'\bDETACH\b',
]


class SQLValidator:
    @staticmethod
    def is_safe(sql: str) -> tuple[bool, str]:
        sql_upper = sql.upper()
        sql_stripped = sql.strip()

        if not (sql_stripped.upper().startswith('SELECT') or
                sql_stripped.upper().startswith('WITH')):
            return False, "Only SELECT queries are allowed"

        for keyword in FORBIDDEN_KEYWORDS:
            pattern = r'\b' + keyword + r'\b'
            if re.search(pattern, sql_upper):
                return False, f"Forbidden keyword detected: {keyword}"

        for pattern in DANGEROUS_PATTERNS:
            if re.search(pattern, sql_upper):
                return False, "Potentially dangerous SQL pattern detected"

        sql_no_trailing = sql_stripped.rstrip(';')
        if ';' in sql_no_trailing:
            return False, "Multiple SQL statements not allowed"

        if '--' in sql or '/*' in sql:
            return False, "SQL comments not allowed"

        if re.search(r'0x[0-9a-fA-F]+', sql):
            return False, "Hex-encoded values not allowed"

        if re.search(r'\bCHAR\s*\(', sql_upper):
            return False, "CHAR() function not allowed"

        return True, ""

    @staticmethod
    def add_limit_if_needed(sql: str) -> str:
        sql_upper = sql.upper()

        if 'LIMIT' in sql_upper:
            return sql

        if any(keyword in sql_upper for keyword in ['GROUP BY', 'COUNT(', 'SUM(', 'AVG(', 'MAX(', 'MIN(']):
            return sql

        sql = sql.rstrip(';')
        return f"{sql} LIMIT {DEFAULT_LIMIT};"

    @staticmethod
    def validate(sql: str) -> tuple[bool, str]:
        is_safe, error = SQLValidator.is_safe(sql)
        if not is_safe:
            return False, error

        sql = SQLValidator.add_limit_if_needed(sql)
        return True, sql

File: lib/sql-validator.ts
import { DEFAULT_LIMIT, FORBIDDEN_KEYWORDS } from './constants';

const DANGEROUS_PATTERNS = [
    /\bEXEC\b/i, /\bEXECUTE\b/i,
    /\bxp_\w+/i, /\bsp_\w+/i,
    /\bINTO\s+OUTFILE\b/i,
    /\bINTO\s+DUMPFILE\b/i,
    /\bLOAD_FILE\b/i,
    /\bUNION\s+SELECT\b/i,
    /\bSLEEP\s*\(/i,
    /\bBENCHMARK\s*\(/i,
    /\bWAITFOR\b/i,
    /\bPG_SLEEP\b/i,
    /\bATTACH\b/i,
    /\bDETACH\b/i,
];

export class SQLValidator {
    static isSafe(sql: string): { isSafe: boolean; error: string } {
        const sqlUpper = sql.toUpperCase();
        const sqlStripped = sql.trim();

        if (!sqlStripped.toUpperCase().startsWith('SELECT') &&
            !sqlStripped.toUpperCase().startsWith('WITH')) {
            return { isSafe: false, error: "Only SELECT queries are allowed" };
        }

        for (const keyword of FORBIDDEN_KEYWORDS) {
            // Word boundary check
            const pattern = new RegExp(`\\b${keyword}\\b`, 'i');
            if (pattern.test(sqlUpper)) {
                return { isSafe: false, error: `Forbidden keyword detected: ${keyword}` };
            }
        }

        for (const pattern of DANGEROUS_PATTERNS) {
            if (pattern.test(sqlUpper)) {
                return { isSafe: false, error: "Potentially dangerous SQL pattern detected" };
            }
        }

        const sqlNoTrailing = sqlStripped.replace(/;+$/, '');
        if (sqlNoTrailing.includes(';')) {
            return { isSafe: false, error: "Multiple SQL statements not allowed" };
        }

        if (sql.includes('--') || sql.includes('/*')) {
            return { isSafe: false, error: "SQL comments not allowed" };
        }

        if (/0x[0-9a-fA-F]+/.test(sql)) {
            return { isSafe: false, error: "Hex-encoded values not allowed" };
        }

        if (/\bCHAR\s*\(/.test(sqlUpper)) {
            return { isSafe: false, error: "CHAR() function not allowed" };
        }

        return { isSafe: true, error: "" };
    }

    static addLimitIfNeeded(sql: string): string {
        const sqlUpper = sql.toUpperCase();

        if (sqlUpper.includes('LIMIT')) {
            return sql;
        }

        // Don't add limit if using aggregates that return few rows usually
        // But check if GROUP BY is present? The Python logic checks these:
        const aggregations = ['GROUP BY', 'COUNT(', 'SUM(', 'AVG(', 'MAX(', 'MIN('];
        if (aggregations.some(keyword => sqlUpper.includes(keyword))) {
            return sql;
        }

        const cleanedSql = sql.trim().replace(/;+$/, '');
        return `${cleanedSql} LIMIT ${DEFAULT_LIMIT};`;
    }

    static validate(sql: string): { isValid: boolean; error: string; sql: string } {
        const { isSafe, error } = SQLValidator.isSafe(sql);
        if (!isSafe) {
            return { isValid: false, error, sql };
        }

        const newSql = SQLValidator.addLimitIfNeeded(sql);
        return { isValid: true, error: "", sql: newSql };
    }
}
